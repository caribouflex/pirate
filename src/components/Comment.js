import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import UserDetails from "./UserDetails";
import { theme } from "../style/theme";
import KidNavigation from "./KidNavigation";

const Container = styled.div`
  color: ${theme.colors.font};
  display: flex;
  flex-direction: column;
  padding: 15px;
  border-bottom: 1px solid ${theme.colors.primary}70;
`;

const CommentText = styled.article`
  padding: 10px 0;
  word-wrap: break-word;
`;

const propTypes = {
  author: PropTypes.string,
  date: PropTypes.number,
  id: PropTypes.number.isRequired,
  loadResponses: PropTypes.func.isRequired,
  parent: PropTypes.number,
  responsesId: PropTypes.arrayOf(PropTypes.number),
  text: PropTypes.string
};

const defaultProps = {
  date: undefined,
  responsesId: [],
  parent: null,
  author: "",
  text: ""
};

const Comment = ({
  id,
  author,
  text,
  date,
  loadResponses,
  responsesId,
  parent
}) => {
  const handleClick = () => {
    loadResponses(responsesId, id, parent);
  };
  return (
    <Container>
      <UserDetails author={author} date={date} />
      <CommentText dangerouslySetInnerHTML={{ __html: text }} />
      <KidNavigation
        disabled={responsesId.length === 0}
        loadKids={handleClick}
        kidsCount={responsesId.length}
        text={responsesId.length > 1 ? " responses" : " response"}
      />
    </Container>
  );
};

Comment.propTypes = propTypes;
Comment.defaultProps = defaultProps;

export default Comment;
//used for tests
export { CommentText };
