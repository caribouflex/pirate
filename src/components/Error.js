import styled from "styled-components";

const Error = styled.div`
  text-align: center;
  color: #ee3123;
  margin: 20px;
  font-size: 18px;
`;

export default Error;
