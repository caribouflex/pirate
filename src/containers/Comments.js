import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Comment from "../components/Comment";
import { getComments } from "../redux-saga/actions/action";

const propTypes = {
  comments: PropTypes.shape({})
};

const defaultProps = {
  comments: {}
};

const Comments = ({ comments, getCommentsAction }) => {
  return (
    <div>
      {Object.keys(comments).map(id => {
        const comment = comments[id];
        return (
          <div key={comment.id}>
            <Comment
              author={comment.by}
              text={comment.text}
              date={comment.time}
              responseCount={comment.kids && comment.kids.length}
            />
            <button onClick={() => getCommentsAction(comment.kids, comment.id)}>
              Load comment childs
            </button>
          </div>
        );
      })}
    </div>
  );
};

Comments.propTypes = propTypes;
Comments.defaultProps = defaultProps;

const mapStateToProps = state => {
  return {
    comments: state.comments[state.comments.selectedStoryId]
  };
};

const mapDispatchToProps = {
  getCommentsAction: getComments
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Comments);
