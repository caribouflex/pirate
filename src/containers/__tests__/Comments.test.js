import React from "react";
import { shallow } from "enzyme";
import { Comments } from "../Comments";
import Error from "../../components/Error";
import Comment from "../../components/Comment";

describe("<Comments />", () => {
  let wrapper;
  const getCommentsAction = jest.fn();
  const setSelectedCommentAction = jest.fn();
  const setSelectedStoryAction = jest.fn();

  const mockComments = {
    19630105: {
      by: "dmethvin",
      id: 19630105,
      kids: [
        19630188,
        19630645,
        19630399,
        19630134,
        19630111,
        19630131,
        19630233
      ],
      parent: 19629614,
      text:
        "AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two.",
      time: 1554943687,
      type: "comment"
    }
  };

  it("Should renders without crashing", () => {
    const wrapper = shallow(
      <Comments
        comments={mockComments}
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    expect(wrapper.exists()).toBe(true);
  });

  it("Should renders with mock data", () => {
    wrapper = shallow(
      <Comments
        comments={mockComments}
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should renders 1 Comment", () => {
    wrapper = shallow(
      <Comments
        comments={mockComments}
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    expect(wrapper.find(Comment).dive().length).toBe(1);
  });

  it("Should be empty with no mock values", () => {
    wrapper = shallow(
      <Comments
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should renders the <Error/> components", () => {
    const wrapperError = shallow(
      <Comments
        errorMessage="An Error Occured"
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    expect(wrapperError.find(Error).length).toBe(1);
  });

  it("Should not call getCommentsAction if id exists", () => {
    const commentsId = [123456];
    const wrapper = shallow(
      <Comments
        allCommentsId={commentsId}
        getCommentsAction={getCommentsAction}
        setSelectedCommentAction={setSelectedCommentAction}
        setSelectedStoryAction={setSelectedStoryAction}
        visible
      />
    );
    const instance = wrapper.instance();
    instance.handleClick([], 123456, 12);
    expect(getCommentsAction).not.toHaveBeenCalled();
  });
});
