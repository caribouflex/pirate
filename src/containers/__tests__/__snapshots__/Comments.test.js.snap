// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments /> Should be empty with no mock values 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Comments
    allCommentsId={Array []}
    comments={Object {}}
    errorMessage={null}
    getCommentsAction={[MockFunction]}
    parentId={-1}
    setSelectedCommentAction={[MockFunction]}
    setSelectedStoryAction={[MockFunction]}
    visible={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <UNDEFINED
          onBackClick={[Function]}
          onCloseClick={[Function]}
          parentId={-1}
        />,
        Array [],
      ],
      "visible": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onBackClick": [Function],
          "onCloseClick": [Function],
          "parentId": -1,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-VigVT",
        "isStatic": false,
        "rules": Array [
          "
  ",
          [Function],
          ";
  width: 50%;
  border-radius: 4px;
  background-color: ",
          "#1f2833",
          ";
  margin: 20px;

  @media (max-width: 1070px) {
    position: absolute;
    margin: 0;
    left: 0;
    top: 80px;
    bottom: 0;
    right: 0;
    width: 100%;
    height: fit-content;
    z-index: 10;
  }
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-VigVT",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UNDEFINED
            onBackClick={[Function]}
            onCloseClick={[Function]}
            parentId={-1}
          />,
          Array [],
        ],
        "visible": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onBackClick": [Function],
            "onCloseClick": [Function],
            "parentId": -1,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-VigVT",
          "isStatic": false,
          "rules": Array [
            "
  ",
            [Function],
            ";
  width: 50%;
  border-radius: 4px;
  background-color: ",
            "#1f2833",
            ";
  margin: 20px;

  @media (max-width: 1070px) {
    position: absolute;
    margin: 0;
    left: 0;
    top: 80px;
    bottom: 0;
    right: 0;
    width: 100%;
    height: fit-content;
    z-index: 10;
  }
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-VigVT",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<Comments /> Should renders with mock data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Comments
    allCommentsId={Array []}
    comments={
      Object {
        "19630105": Object {
          "by": "dmethvin",
          "id": 19630105,
          "kids": Array [
            19630188,
            19630645,
            19630399,
            19630134,
            19630111,
            19630131,
            19630233,
          ],
          "parent": 19629614,
          "text": "AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two.",
          "time": 1554943687,
          "type": "comment",
        },
      }
    }
    errorMessage={null}
    getCommentsAction={[MockFunction]}
    parentId={-1}
    setSelectedCommentAction={[MockFunction]}
    setSelectedStoryAction={[MockFunction]}
    visible={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <UNDEFINED
          onBackClick={[Function]}
          onCloseClick={[Function]}
          parentId={-1}
        />,
        Array [
          <Comment
            author="dmethvin"
            date={1554943687}
            id={19630105}
            loadResponses={[Function]}
            parent={19629614}
            responsesId={
              Array [
                19630188,
                19630645,
                19630399,
                19630134,
                19630111,
                19630131,
                19630233,
              ]
            }
            text="AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two."
          />,
        ],
      ],
      "visible": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onBackClick": [Function],
          "onCloseClick": [Function],
          "parentId": -1,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": [Function],
        },
      },
      Object {
        "instance": null,
        "key": "19630105",
        "nodeType": "function",
        "props": Object {
          "author": "dmethvin",
          "date": 1554943687,
          "id": 19630105,
          "loadResponses": [Function],
          "parent": 19629614,
          "responsesId": Array [
            19630188,
            19630645,
            19630399,
            19630134,
            19630111,
            19630131,
            19630233,
          ],
          "text": "AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two.",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-VigVT",
        "isStatic": false,
        "rules": Array [
          "
  ",
          [Function],
          ";
  width: 50%;
  border-radius: 4px;
  background-color: ",
          "#1f2833",
          ";
  margin: 20px;

  @media (max-width: 1070px) {
    position: absolute;
    margin: 0;
    left: 0;
    top: 80px;
    bottom: 0;
    right: 0;
    width: 100%;
    height: fit-content;
    z-index: 10;
  }
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-VigVT",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UNDEFINED
            onBackClick={[Function]}
            onCloseClick={[Function]}
            parentId={-1}
          />,
          Array [
            <Comment
              author="dmethvin"
              date={1554943687}
              id={19630105}
              loadResponses={[Function]}
              parent={19629614}
              responsesId={
                Array [
                  19630188,
                  19630645,
                  19630399,
                  19630134,
                  19630111,
                  19630131,
                  19630233,
                ]
              }
              text="AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two."
            />,
          ],
        ],
        "visible": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onBackClick": [Function],
            "onCloseClick": [Function],
            "parentId": -1,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": [Function],
          },
        },
        Object {
          "instance": null,
          "key": "19630105",
          "nodeType": "function",
          "props": Object {
            "author": "dmethvin",
            "date": 1554943687,
            "id": 19630105,
            "loadResponses": [Function],
            "parent": 19629614,
            "responsesId": Array [
              19630188,
              19630645,
              19630399,
              19630134,
              19630111,
              19630131,
              19630233,
            ],
            "text": "AMP is just a set of conventions and limitations that, when followed, make for a fast site. Anyone can make a fast site if they follow similar rules. Most sites don&#x27;t do that because either the developers want to use something that&#x27;s &quot;nicer&quot; to code but a lot bigger, or because the marketing department insists on loading 12 different tracking and analytics progams--when they probably only use one or two.",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-VigVT",
          "isStatic": false,
          "rules": Array [
            "
  ",
            [Function],
            ";
  width: 50%;
  border-radius: 4px;
  background-color: ",
            "#1f2833",
            ";
  margin: 20px;

  @media (max-width: 1070px) {
    position: absolute;
    margin: 0;
    left: 0;
    top: 80px;
    bottom: 0;
    right: 0;
    width: 100%;
    height: fit-content;
    z-index: 10;
  }
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-VigVT",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
